### PR TITLE
feat: [CDS-71746]: execution changes (GRAPH + CONSOLE) for retry step group as failure strategy

### PR DESCRIPTION
**Pull request by vikrant-harness**
### Summary

Feature Request :- To support `RetryStepGroup` as failure strategy in execution page , the studio changes for the same were already completed

Summary:-  Whenever we have `RetryStepGroup FS` configured for a step group and any step fails in the step group the step group `internally retries` the execution and the data for the same is received in the polling API call under the `interrupt histories` for the step group , Using the ID's received in the interrupt histories we fetch the `sub-graph of the step group` on demand by the user if the user selects any `previous retry ID` , The configuration for the same is stored in the local storage to maintain across the reloads.

#### Screenshots

https://github.com/harness/harness-core-ui/assets/106532291/9e1605f2-63c3-45d4-8ebf-8f2fb3f16846

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/useRetryStepGroup.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import React from 'react'
```
Comment (PratyushGar) [inline src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/useRetryStepGroup.tsx:1]: Can be renamed to useGetRetryStepGroupData.tsx

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx [hunk 2]
```diff
@@ -179,6 +233,27 @@ export function StepGroupNode(props: any): JSX.Element {
...
+    })
+  }
+
+  function goToCurrentExecution(): void {
+    setRetryStepGroupParams(prevState => {
+      const newRetryParams = { ...prevState, [stepGroupData.data?.baseFqn]: '' }
+      return newRetryParams
+    })
+    updateQueryParams({ step: [] as unknown as string /* this removes the param fro query */ })
+  }
```
Comment (PratyushGar) [inline src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx:249]: Use this from hook

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx [hunk 3]
```diff
@@ -72,6 +81,51 @@ export function StepGroupNode(props: any): JSX.Element {
...
+      accountIdentifier: accountId,
+      projectIdentifier,
+      orgIdentifier
+    },
+    pathParams: {
+      planExecutionId: executionIdentifier,
+      nodeExecutionId: currentStepGroupRetryId
+    },
+    lazy: isEmpty(currentStepGroupRetryId)
+  })
```
Comment (PratyushGar) [inline src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx:117]: You can expose this data from hook and keep updation logic in stepGroupNode with `processExecutionDataV1`

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx [hunk 4]
```diff
@@ -72,6 +81,51 @@ export function StepGroupNode(props: any): JSX.Element {
   const stepsData = stepGroupData?.steps || stepGroupData?.template?.templateInputs?.steps
   const isParentMatrix = defaultTo(props?.isParentMatrix, false)
   const { module, source = 'executions' } = useParams<PipelineType<ExecutionPathProps>>()
+  const [retryStepGroupParams, setRetryStepGroupParams] = useLocalStorage<{ [key: string]: string }>(
+    'retryStepGroup',
+    {}
```
Comment (PratyushGar) [inline src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx:86]: nit: Check if computations can be avoided using isExecutionView boolean, as retry logic is only applicable for execution view

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx [hunk 5]
```diff
@@ -72,6 +81,51 @@ export function StepGroupNode(props: any): JSX.Element {
...
+    pathParams: {
+      planExecutionId: executionIdentifier,
+      nodeExecutionId: currentStepGroupRetryId
+    },
+    lazy: isEmpty(currentStepGroupRetryId)
+  })
+
+  React.useEffect(() => {
+    if (executionNode?.data?.executionGraph) {
+      setRetryStepGroupStepsData(processExecutionDataV1(executionNode?.data?.executionGraph))
```
Comment (PratyushGar) [inline src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx:121]:
```suggestion
      setRetryStepGroupStepsData(processExecutionDataV1(executionNode.data.executionGraph))
```

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx [hunk 6]
```diff
@@ -179,6 +233,27 @@ export function StepGroupNode(props: any): JSX.Element {
...
+    })
+    updateQueryParams({ step: [] as unknown as string /* this removes the param fro query */ })
+  }
+
+  let retryCount = interruptHistories.length
+  if (currentStepGroupRetryId) {
+    retryCount = interruptHistories.findIndex(
+      ({ interruptConfig }) => interruptConfig?.retryInterruptConfig?.retryId === currentStepGroupRetryId
+    )
+  }
```
Comment (PratyushGar) [inline src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx:256]: Avoid inline updates, can be moved to useMemo

Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx]: done

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx (hunk 5 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx]: done

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx (hunk 4 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx]: will do in next pr

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx (hunk 3 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx]: done

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx (hunk 2 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx]: done

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/useRetryStepGroup.tsx (hunk 1 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/useRetryStepGroup.tsx]: done

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupGraph/StepGroupGraph.tsx [hunk 7]
```diff
@@ -174,29 +175,27 @@ function StepGroupGraph(props: StepGroupGraphProps): React.ReactElement {
...
-                  entityType: DiagramType.CreateNew,
-                  node: props,
-                  relativeBasePath: `${relativeBasePath}.steps`
-                }
-              })
-            }}
-            name={null}
-          />
-        )
+        <ContainerSpinner padding={'small'} style={{ height: '64px', width: '64px' }} message="Loading..." />
```
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupGraph/StepGroupGraph.tsx]: done

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx [hunk 8]
```diff
@@ -72,6 +80,38 @@ export function StepGroupNode(props: any): JSX.Element {
...
+    return getInterruptHistoriesFromType(stepGroupData?.data?.interruptHistories, Interrupt.RETRY)
+  }, [stepGroupData?.data?.interruptHistories])
+
+  const { executionNode, retryStepGroupParams, loading, goToCurrentExecution, goToRetryStepExecution } =
+    useGetRetryStepGroupData({
+      currentStepGroupRetryId
+    })
+
+  React.useEffect(() => {
+    let currentStepGrpRetryId = retryStepGroupParams[`${stepGroupData.data?.baseFqn}`]
```
Comment (PratyushGar) [inline src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx:94]: UseEffect and `currentStepGroupRetryId` state can be removed and useMemo can be used

**Review by PratyushGar — APPROVED** [review]

File: src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx (hunk 8 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineDiagram/Nodes/StepGroupNode/StepGroupNode.tsx:94]: The component re-render wasn't getting triggered so needed useEffect to fetch latest data

**Review by bhavya-sinha — APPROVED** [review]

**Pull request merged**